File: python/examples/triangle_window.py
#!/usr/bin/env python3
"""Triangle example — animated colored triangle in a window via TaskGraph.

Offscreen RenderTarget -> render_pass -> copy_render_target_to_swapchain -> present.

Usage:
    python triangle_window.py
"""

from __future__ import annotations

import math
import sys

import glfw
import goldy
import numpy as np


def make_scene_rt(device: goldy.Device, surface: goldy.Surface) -> goldy.RenderTarget:
    width = max(surface.width, 1)
    height = max(surface.height, 1)
    return goldy.RenderTarget(device, width, height, surface.format)


def main() -> int:
    print("Goldy Python Triangle Window (TaskGraph)")
    print("=" * 40)
    print("Press Escape or close the window to exit\n")

    if not glfw.init():
        print("Failed to initialize GLFW", file=sys.stderr)
        return 1

    glfw.window_hint(glfw.CLIENT_API, glfw.NO_API)
    window = glfw.create_window(800, 600, "Goldy - Animated Triangle (Python)", None, None)
    if not window:
        glfw.terminate()
        print("Failed to create GLFW window", file=sys.stderr)
        return 1

    instance = goldy.Instance()
    device = instance.request_adapter().request_device()
    surface = goldy.Surface.from_glfw(device, window)

    shader = goldy.ShaderModule.from_slang(device, goldy.Builtins.VERTEX_COLOR_2D)
    pipeline = goldy.RenderPipeline(
        device,
        shader,
        shader,
        goldy.RenderPipelineDesc(
            vertex_layout=goldy.VertexBufferLayout.vertex_2d(),
            target_format=surface.format,
        ),
    )

    vertices = np.array(
        [
            0.0,
            -0.5,
            1.0,
            0.0,
            0.0,
            1.0,
            -0.5,
            0.5,
            0.0,
            1.0,
            0.0,
            1.0,
            0.5,
            0.5,
            0.0,
            0.0,
            1.0,
            1.0,
        ],
        dtype=np.float32,
    )
    vertex_buffer = goldy.Buffer(device, vertices, goldy.BufferKind.SCATTERED)

    scene_rt = make_scene_rt(device, surface)
    frame_graph = goldy.TaskGraph()
    frame_count = 0

    try:
        while not glfw.window_should_close(window):
            fb_width, fb_height = glfw.get_framebuffer_size(window)
            if fb_width > 0 and fb_height > 0:
                if fb_width != surface.width or fb_height != surface.height:
                    surface.resize(fb_width, fb_height)
                    scene_rt = make_scene_rt(device, surface)

            t = math.sin(frame_count * 0.02) * 0.5 + 0.5
            bg = goldy.Color(0.1 + t * 0.1, 0.1 + t * 0.05, 0.2 + t * 0.1, 1.0)

            frame_graph.clear()
            with frame_graph.render_pass("triangle", scene_rt) as rp:
                (
                    rp.bind_buffer(vertex_buffer, goldy.NodeAccess.READ)
                    .clear(bg)
                    .set_pipeline(pipeline)
                    .set_vertex_buffer(0, vertex_buffer)
                    .draw(range(3))
                )

            swapchain = frame_graph.declare_swapchain_output()
            frame_graph.copy_render_target_to_swapchain(scene_rt, swapchain)

            frame = surface.acquire()
            surface.submit_graph_to_frame(frame_graph, frame)
            surface.present(frame)

            frame_count += 1
            glfw.poll_events()

            if glfw.get_key(window, glfw.KEY_ESCAPE) == glfw.PRESS:
                break
    finally:
        glfw.destroy_window(window)
        glfw.terminate()

    print("Done!")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: dotnet/Goldy.Examples/TriangleWindow.cs
using System.Runtime.InteropServices;
using Goldy;
using Silk.NET.GLFW;

namespace Goldy.Examples;

/// <summary>
/// Windowed triangle via GLFW + retained Scheme (offscreen RT → bind_render_target).
/// </summary>
static class TriangleWindow
{
    [StructLayout(LayoutKind.Sequential)]
    struct Vertex2D
    {
        public float Px, Py;
        public float R, G, B, A;
    }

    static (SchemeRenderTargetLease rt, Transaction transaction) RecordScheme(
        Scheme scheme,
        SurfaceExchange surface,
        RenderPipeline pipeline,
        Parcel vertexParcel,
        Color bg)
    {
        var rt = scheme.LeaseRenderTarget(
            Math.Max(surface.Width, 1u),
            Math.Max(surface.Height, 1u),
            surface.Format);
        using (var pass = scheme.RenderPass("triangle", rt))
        {
            pass
                .WithParcel(vertexParcel, NodeAccess.Read)
                .Clear(bg)
                .SetPipeline(pipeline)
                .SetVertexBuffer(0, vertexParcel)
                .Draw(3);
        }
        var transaction = surface.BindRenderTarget(scheme, rt);
        return (rt, transaction);
    }

    public static unsafe void Run()
    {
        Console.WriteLine("Goldy .NET Triangle Window (Scheme + SurfaceExchange)");
        Console.WriteLine(new string('=', 40));
        Console.WriteLine("Press Escape or close the window to exit\n");

        var glfw = Glfw.GetApi();
        if (!glfw.Init())
            throw new InvalidOperationException("glfwInit failed");

        glfw.WindowHint(WindowHintClientApi.ClientApi, ClientApi.NoApi);
        var window = glfw.CreateWindow(800, 600, "Goldy - Triangle (.NET / Scheme)", null, null);
        if (window == null)
        {
            glfw.Terminate();
            throw new InvalidOperationException("glfwCreateWindow failed");
        }

        try
        {
            using var instance = new Instance();
            using var device = instance.RequestAdapter().RequestDevice();
            using var ctx = device.CreateContext();
            using var surface = GlfwSurfaceExchange.Create(ctx, window);

            using var shader = new ShaderModule(device, ShaderModule.BuiltinVertexColor2D);
            using var pipeline = new RenderPipeline(
                device,
                shader,
                shader,
                new RenderPipelineDesc
                {
                    VertexAttributes = VertexLayouts.Vertex2D,
                    VertexStride = 24,
                    TargetFormat = surface.Format,
                });

            ReadOnlySpan<Vertex2D> vertices =
            [
                new() { Px = 0.0f, Py = -0.5f, R = 1, G = 0, B = 0, A = 1 },
                new() { Px = -0.5f, Py = 0.5f, R = 0, G = 1, B = 0, A = 1 },
                new() { Px = 0.5f, Py = 0.5f, R = 0, G = 0, B = 1, A = 1 },
            ];
            using var retainedPool = new RetainedPool(device);
            using var vertexBuffer = retainedPool.AcquireBuffer(vertices, BufferKind.Scattered);
            using var vertexParcel = vertexBuffer.Field(0);

            var bg = new Color(0.1f, 0.1f, 0.2f, 1.0f);
            var scheme = new Scheme(ctx);
            var (sceneRt, transaction) = RecordScheme(scheme, surface, pipeline, vertexParcel, bg);

            while (!glfw.WindowShouldClose(window))
            {
                glfw.GetFramebufferSize(window, out int fbWidth, out int fbHeight);
                if (fbWidth > 0 && fbHeight > 0)
                {
                    var w = (uint)fbWidth;
                    var h = (uint)fbHeight;
                    if (w != surface.Width || h != surface.Height)
                    {
                        surface.Resize(w, h);
                        sceneRt.Dispose();
                        transaction.Dispose();
                        scheme.Dispose();
                        scheme = new Scheme(ctx);
                        (sceneRt, transaction) = RecordScheme(scheme, surface, pipeline, vertexParcel, bg);
                    }
                }

                using var submission = scheme.Submit();
                using var claim = transaction.Claim(submission);
                claim.Consume();

                glfw.PollEvents();

                if (glfw.GetKey(window, Keys.Escape) == (int)InputAction.Press)
                    glfw.SetWindowShouldClose(window, true);
            }

            sceneRt.Dispose();
            transaction.Dispose();
            scheme.Dispose();
            Console.WriteLine("Done!");
        }
        finally
        {
            glfw.DestroyWindow(window);
            glfw.Terminate();
        }
    }
}
